saleor/graphql/attribute/mutations/attribute_bulk_update.py

```diff
@@ -0,0 +1,636 @@
+            )
+            attribute_data["fields"]["remove_values"] = cleaned_remove_values
+
+        if add_values:
+            cleaned_add_values = clean_values(
+                add_values,
+                attr.input_type,
+                values_existing_external_refs,
+                duplicated_values_external_ref,
+                attribute_index,
+                index_error_map,
+                attr,
+                "addValues",
+                AttributeBulkUpdateError,
+            )
+            attribute_data["fields"]["add_values"] = cleaned_add_values
+
+        return attribute_data
+
+    @classmethod
+    def clean_remove_values(
+        cls, remove_values, attribute, attribute_index, index_error_map
+    ):
+        clean_remove_values = []
+
+        for index, value_global_id in enumerate(remove_values):
+            try:
+                obj_type, value_db_id = from_global_id_or_error(
+                    value_global_id, only_type="AttributeValue", raise_error=True
+                )
+            except GraphQLError as exc:
+                index_error_map[attribute_index].append(
+                    AttributeBulkUpdateError(
+                        path=f"removeValues.{index}",
+                        message=str(exc),
+                        code=AttributeBulkUpdateErrorCode.INVALID.value,
+                    )
+                )
+                continue
+
+            # values are prefetched
+            values = attribute.values.all()
+            value = next((obj for obj in values if str(obj.pk) == value_db_id), None)
+
+            if not value:
+                msg = f"Value {value_global_id} does not belong to this attribute."
+                index_error_map[attribute_index].append(
+                    AttributeBulkUpdateError(
+                        path=f"removeValues.{index}",
+                        message=msg,
+                        code=AttributeBulkUpdateErrorCode.INVALID.value,
+                    )
+                )
+            else:
+                clean_remove_values.append(value)
+        return clean_remove_values
+
+    @classmethod
+    def update_attributes(
+        cls,
+        info: ResolveInfo,
+        cleaned_inputs_map: dict[int, dict],
+        error_policy: str,
+        index_error_map: dict[int, List[AttributeBulkUpdateError]],
+    ) -> List[dict]:
+        instances_data_and_errors_list: List[dict] = []
+
+        for index, cleaned_input in cleaned_inputs_map.items():
+            if not cleaned_input:
+                instances_data_and_errors_list.append(
+                    {"instance": None, "errors": index_error_map[index]}
+                )
+                continue
+
+            attr = cleaned_input["instance"]
+            fields = cleaned_input["fields"]
+            remove_values = fields.pop("remove_values", [])
+            add_values = fields.pop("add_values", [])
+
+            try:
```

Comment:
What is a part of the data that can raise an exception here?  methods called in `if fields:`? If yes, why are wrapping everything in the try-except?
The try: except is quite big. Do we need to create a `data` structure inside it? 

What do you think about this one 👇 
```python
            if fields:
                try:
                    attr = cls.construct_instance(attr, fields)
                    cls.clean_instance(info, attr)
                except ValidationError as exc:
                    for key, errors in exc.error_dict.items():
                        for e in errors:
                            index_error_map[index].append(
                                AttributeBulkUpdateError(
                                    path=to_camel_case(key),
                                    message=e.messages[0],
                                    code=e.code,
                                )
                            )
                    instances_data_and_errors_list.append(
                        {"instance": None, "errors": index_error_map[index]}
                    )
                    continue

            data = {
                "instance": attr,
                "errors": index_error_map[index],
                "remove_values": remove_values,
                "add_values": [],
                "attribute_updated": True if fields else False,
            }
```
The scope of try-except is small. the `if fields` gives you immediately context that this exception can be raised only when `if fields`. The `data` is outside as it is not part of try-except block

Reply:
Good idea thanks!